Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift

```diff
@@ -301,61 +280,52 @@ public final class EventLoopConnectionPool<Source> where Source: ConnectionPoolS
-            knownConnection.lastUsed = Date()
-            self.available.append(knownConnection)
-        } else {
-            //this is connection that we don't know yet, it was returned by a future and the client just wants to release it.
-            let newConnection = PrunableConnection(connection: connection, lastUsed: Date())
-            self.available.append(newConnection)
-        }
+        // Push the connection onto the end of the available list so it's the first one to get used
+        // on the next request.
+        logger.trace("Releasing pool connection on \(self.eventLoop.description), \(self.available.count + (connection.isClosed ? 0 : 1)) connections available.")
 
-        // now that we know a new connection is available, we should
-        // take this chance to fulfill one of the waiters
-        let waiter: WaitlistItem?
-        if !self.waiters.isEmpty {
-            waiter = self.waiters.removeFirst().value
+        if var existing = self.available.first(where: { $0 === connection }) {
+            existing.lastUsed = Date()
+            self.available.append(existing)
         } else {
-            waiter = nil
+            // This is a connection that we don't know yet, it was returned by a future and the client just wants to release it.
+            self.available.append(.init(connection: connection, lastUsed: Date()))
         }
 
-        // if there is a waiter, request a connection for it
-        if let (logger, promise, _) = waiter {
-            logger.debug("Fulfilling connection waitlist request")
-            self.requestConnection(
-                logger: logger
-            ).cascade(to: promise)
+        // For as long as there are connections available, try to dequeue waiters. Even if the available
+        // connection(s) are closed, the request logic will try to open new ones.
+        while !self.available.isEmpty, !self.waiters.isEmpty {
+            let waiter = self.waiters.removeFirst()
+
+            logger.debug("Servicing connection waitlist item with id \(waiter.key)")
+            waiter.value.timeoutTask.cancel()
+            self._requestConnection0(logger: waiter.value.logger).cascade(to: waiter.value.promise)
         }
     }
-    
-    
-    private func pruneConnections() {
-        // dispatch to event loop thread if necessary
-        guard self.eventLoop.inEventLoop else {
-            return self.eventLoop.execute {
-                self.pruneConnections()
-            }
-        }
 
-        self.lock.withLockVoid {
-            self.available.filter{!$0.originalConnection.isClosed}.forEach { conn in
-                if Date().timeIntervalSince(conn.lastUsed) >= self.maxIdleTimeBeforePrunning {
+    func setupPruneConnectionTask(
+        pruneInterval: TimeAmount,
+        maxIdleTimeBeforePruning: TimeAmount
+    ) {
+        let maxIdleTimeBeforePruning = Double(maxIdleTimeBeforePruning.nanoseconds) / 1_000_000_000
+        self.eventLoop.scheduleRepeatedTask(
```

Comment:
I don't think there is a need. There is already a `task.cancel()` when the class is deinitialized.

If i still should store it, where do you think it would be useful?

Reply:
That's fine I guess, it just reads like vulnerable code should someone edit this down the line. Easy to make a mistake here